### PR TITLE
[ST-534] Fix reverting logic and unskip tests

### PR DESCRIPTION
**Pull request by anikitinDSR**
Signed-off-by: Andrew Nikitin <andrew.nikitin@dsr-corporation.com>

### CODE REVIEW

File: sovtokenfees/sovtokenfees/static_fee_req_handler.py [hunk 1]
```diff
@@ -186,12 +189,26 @@ def post_batch_created(self, ledger_id, state_root):
                                            self.token_ledger.uncommitted_size)
 
     def post_batch_rejected(self, ledger_id):
-        count_reverted = TokenReqHandler.on_batch_rejected(self.utxo_cache, self.token_tracker, self.token_state, self.token_ledger)
-        self.fee_txns_in_current_batch = 0
+        if ledger_id == TOKEN_LEDGER_ID:
+            # TODO: Need to improve this logic for case, when we got a XFER txn with fees
+            # All of other txn with fees it's a 2 steps, "apply txn" and "apply fees"
+            # But for XFER txn with fees we do only "apply fees with transfer too"
+            self.token_tracker.reject_batch()
```
Comment (ashcherbakov) [inline sovtokenfees/sovtokenfees/static_fee_req_handler.py]: I suggest to move all `token_tracker` related logic to StaticFeesReqHandler. TokenReqHandler should take care of utxo cache only (the same way as DomainReqHandler takes care of idr cache)
So, every StaticFeesReqHandler's method (batch created/rejected/committed) should look like this:
- if this is payment ledger - return
- apply/reject/commit to the token_tracker
- do other logic based on token_tracker

File: sovtoken/sovtoken/token_req_handler.py [hunk 2]
```diff
@@ -188,13 +188,12 @@ def _add_new_output(self, output: Output, is_committed=False):
                             is_committed=is_committed)
 
     def onBatchCreated(self, state_root, txn_time):
-        self.on_batch_created(self.utxo_cache, self.tracker, self.ledger, state_root)
+        self.on_batch_created(self.utxo_cache, state_root)
```
Comment (ashcherbakov) [inline sovtoken/sovtoken/token_req_handler.py]: Can we remove `self.tracker` instance?

File: sovtokenfees/sovtokenfees/static_fee_req_handler.py [hunk 3]
```diff
@@ -174,24 +174,40 @@ def get_fees(self, request: Request):
...
             self.fee_txns_in_current_batch = 0
-        else:
-            self.token_tracker.apply_batch(self.token_state.headHash,
-                                           self.token_ledger.uncommitted_root_hash,
-                                           self.token_ledger.uncommitted_size)
 
     def post_batch_rejected(self, ledger_id):
-        count_reverted = TokenReqHandler.on_batch_rejected(self.utxo_cache, self.token_tracker, self.token_state, self.token_ledger)
-        self.fee_txns_in_current_batch = 0
+        if ledger_id == TOKEN_LEDGER_ID:
```
Comment (ashcherbakov) [inline sovtokenfees/sovtokenfees/static_fee_req_handler.py]: Should we call `self.token_tracker.reject_batch()` at the beginning of the call?

File: sovtokenfees/sovtokenfees/static_fee_req_handler.py [hunk 4]
```diff
@@ -174,24 +174,40 @@ def get_fees(self, request: Request):
...
+                self.token_state.headHash == uncommitted_hash:
+            return 0
+        self.token_state.revertToHead(uncommitted_hash)
+        self.token_ledger.discardTxns(txn_count)
+        count_reverted = TokenReqHandler.on_batch_rejected(self.utxo_cache)
         logger.debug("Reverted {} txns with fees".format(count_reverted))
 
     def post_batch_committed(self, ledger_id, pp_time, committed_txns,
                              state_root, txn_root):
+        # All changes will be tracked on TokenReqHandler side
```
Comment (ashcherbakov) [inline sovtokenfees/sovtokenfees/static_fee_req_handler.py]: Should we call `self.token_tracker.commit_batch()` at the beginning of the call?

File: sovtokenfees/sovtokenfees/static_fee_req_handler.py [hunk 5]
```diff
@@ -174,43 +178,55 @@ def get_fees(self, request: Request):
...
     def post_batch_rejected(self, ledger_id):
-        count_reverted = TokenReqHandler.on_batch_rejected(self.utxo_cache, self.token_tracker, self.token_state, self.token_ledger)
-        self.fee_txns_in_current_batch = 0
+        uncommitted_hash, uncommitted_txn_root, txn_count = self.token_tracker.reject_batch()
+        if ledger_id == TOKEN_LEDGER_ID:
+            # TODO: Need to improve this logic for case, when we got a XFER txn with fees
+            # All of other txn with fees it's a 2 steps, "apply txn" and "apply fees"
+            # But for XFER txn with fees we do only "apply fees with transfer too"
+            return
+        if txn_count == 0 or self.token_ledger.uncommitted_root_hash == uncommitted_txn_root or \
```
Comment (ashcherbakov) [inline sovtokenfees/sovtokenfees/static_fee_req_handler.py]: Do we really need this check anymore?

**Comment by cam-parra** [comment]
@ashcherbakov looks like a test is breaking here. Can we fix this? I will merge once that is complete

**Comment by ashcherbakov** [comment]
I think we need to do the following to make the logic cleaner:
- make utxo-cache-related logic in TokenReqHandler only (the same way as idr_cache is handled in DomainReqHandler)
- move all tracker-related logic to StaticFeesReqHandler only
- do not call anything (except tracker-related logic) in StaticFeesReqHandler  if the ledger is a payment one (since XFER and FEEs are combined into one txn if we have FEEs set for XFER)
- correctly update last_committed in StaticFeesReqHandler's tracker when catchup is finished. It may require a new hook. Once pluggable requests handlers are implemented, this should not require a hood anymore.

File: sovtoken/sovtoken/main.py [hunk 6]
```diff
@@ -35,7 +28,7 @@ def postTxnAddedToLedgerClbk(node, *args, **kwargs):
         node.ledger_ids.append(TOKEN_LEDGER_ID)
 
     node.ledgerManager.addLedger(TOKEN_LEDGER_ID, ledger,
-                                 postTxnAddedToLedgerClbk=functools.partial(postTxnAddedToLedgerClbk, node))
+                                 postTxnAddedToLedgerClbk=node.postTxnFromCatchupAddedToLedger)
```
Comment (KitHat) [inline sovtoken/sovtoken/main.py]: What does `Clbk` mean?

Comment (anikitinDSR) [reply sovtoken/sovtoken/main.py]: this callback will be called for each txn from Catchup Reply. In this case we will do generic node's logic, like set to state and add to ledger. But for StaticFeesReqHandler i add another callback on ledger's caughtup complete for setting last committed into tracker.

**Pull request merged**